plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergBranching.java

```diff
@@ -0,0 +1,231 @@
+        metastore = ((IcebergConnector) queryRunner.getCoordinator().getConnector(ICEBERG_CATALOG)).getInjector()
+                .getInstance(HiveMetastoreFactory.class)
+                .createMetastore(Optional.empty());
+        fileSystemFactory = getFileSystemFactory(queryRunner);
+
+        return queryRunner;
+    }
+
+    @Test
+    void testCreateBranch()
+    {
+        try (TestTable table = new TestTable(getQueryRunner()::execute, "test_create_branch", "(x int)")) {
+            assertBranch(table.getName(), "main");
+
+            createBranch(table.getName(), "test-branch");
+            assertBranch(table.getName(), "main", "test-branch");
+
+            createBranch(table.getName(), "TEST-BRANCH");
+            assertBranch(table.getName(), "main", "test-branch", "TEST-BRANCH");
+
+            assertQueryFails("ALTER TABLE " + table.getName() + " EXECUTE create_branch('test-branch')", "Branch 'test-branch' already exists");
+        }
+    }
+
+    @Test
+    void testDropBranch()
+    {
+        try (TestTable table = new TestTable(getQueryRunner()::execute, "test_drop_branch", "(x int)")) {
+            assertBranch(table.getName(), "main");
+
+            createBranch(table.getName(), "test-branch");
+            createBranch(table.getName(), "TEST-BRANCH");
+            assertBranch(table.getName(), "main", "test-branch", "TEST-BRANCH");
+
+            dropBranch(table.getName(), "test-branch");
+            dropBranch(table.getName(), "TEST-BRANCH");
+            assertBranch(table.getName(), "main");
+
+            assertQueryFails("ALTER TABLE " + table.getName() + " EXECUTE drop_branch('test-branch')", "Branch 'test-branch' does not exit");
+            assertQueryFails("ALTER TABLE " + table.getName() + " EXECUTE drop_branch('main')", "Cannot drop 'main' branch");
+        }
+    }
+
+    @Test
+    void testFastForward()
+    {
+        try (TestTable table = new TestTable(getQueryRunner()::execute, "test_fast_forward", "(x int)")) {
+            assertBranch(table.getName(), "main");
+
+            createBranch(table.getName(), "test-branch");
+            assertUpdate("INSERT INTO " + table.getName() + " WITH (target_branch = 'test-branch') VALUES 1, 2, 3", 3);
+            assertThat(computeScalar("SELECT count(*) FROM " + table.getName()))
+                    .isEqualTo(0L);
+            assertThat(computeScalar("SELECT count(*) FROM " + table.getName() + " FOR VERSION AS OF 'test-branch'"))
+                    .isEqualTo(3L);
+
+            fastForward(table.getName(), "main", "test-branch");
+            assertThat(computeScalar("SELECT count(*) FROM " + table.getName()))
+                    .isEqualTo(3L);
+            assertThat(computeScalar("SELECT count(*) FROM " + table.getName() + " FOR VERSION AS OF 'test-branch'"))
+                    .isEqualTo(3L);
+
+            // Iceberg creates a new branch if "from" branch does not exist
+            fastForward(table.getName(), "new-branch", "test-branch");
```

Comment:
What should be expected outcome if we tried to fast forwad to existing not related branch (not main)